camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/ConfigInheritanceYamlTest.java

```diff
@@ -300,14 +338,212 @@ public void testOverridesParentConfig() throws Exception {
+                    "  brooklyn.children:",
+                    "  - type: entity-with-keys");
+            
+            Entity app = createStartWaitAndLogApplication(new StringReader(yaml));
+            TestEntity entity = (TestEntity) Iterables.getOnlyElement(app.getChildren());
+    
+            assertEquals(entity.config().get(entity.getEntityType().getConfigKey("map.type-merged")), ImmutableMap.of("myDefaultKey", "myDefaultVal"));
+            assertEquals(entity.config().get(entity.getEntityType().getConfigKey("map.type-always")), ImmutableMap.of("myDefaultKey", "myDefaultVal"));
+            assertEquals(entity.config().get(entity.getEntityType().getConfigKey("map.type-never")), ImmutableMap.of("myDefaultKey", "myDefaultVal"));
+        }
+
+        // Test override defaults in parent entity
+        {
+            String yaml = Joiner.on("\n").join(
+                    "services:",
+                    "- type: org.apache.brooklyn.entity.stock.BasicApplication",
+                    "  brooklyn.config:",
+                    "    map.type-merged:",
+                    "      mykey: myval",
+                    "    map.type-always:",
+                    "      mykey: myval",
+                    "    map.type-never:",
+                    "      mykey: myval",
+                    "  brooklyn.children:",
+                    "  - type: entity-with-keys");
+            
+            Entity app = createStartWaitAndLogApplication(new StringReader(yaml));
+            TestEntity entity = (TestEntity) Iterables.getOnlyElement(app.getChildren());
+    
+            assertEquals(entity.config().get(entity.getEntityType().getConfigKey("map.type-merged")), ImmutableMap.of("mykey", "myval"));
+            assertEquals(entity.config().get(entity.getEntityType().getConfigKey("map.type-always")), ImmutableMap.of("mykey", "myval"));
+            assertEquals(entity.config().get(entity.getEntityType().getConfigKey("map.type-never")), ImmutableMap.of("myDefaultKey", "myDefaultVal"));
+        }
+
+        // Test merging/overriding of explicit config
+        {
+            String yaml = Joiner.on("\n").join(
+                    "services:",
+                    "- type: org.apache.brooklyn.entity.stock.BasicApplication",
+                    "  brooklyn.config:",
+                    "    map.type-merged:",
+                    "      mykey: myval",
+                    "    map.type-always:",
+                    "      mykey: myval",
+                    "    map.type-never:",
+                    "      mykey: myval",
+                    "  brooklyn.children:",
+                    "  - type: entity-with-keys",
+                    "    brooklyn.config:",
+                    "      map.type-merged:",
+                    "        mykey2: myval2",
+                    "      map.type-always:",
+                    "        mykey2: myval2",
+                    "      map.type-never:",
+                    "        mykey2: myval2");
```

Comment:
If 3 levels within a map, then if merging it will do a "deep merge", as per `CollectionMerger`. I agree it's not clear what a user would expect. I suspect some users would think it will be shallow, and other users will think deep (depending on the particularly values in the config map, presumably).

Or do you mean 3 levels of entity (e.g. A extends B, and C extends B)? Then the rules apply at each level: B will merge or override the values of A, and then C will merge or override those values of B (based on the config key definition).